OCMF-en.md

```diff
@@ -709,7 +709,7 @@ provided for the data area.
 | ECDSA-secp192k1-SHA256                               | ECDSA               | secp192k1                                                   | 192 bit    | SHA-256        | 48           |
 | ECDSA-secp256k1-SHA256                               | ECDSA               | secp256k1                                                   | 256 bit    | SHA-256        | 64           |
 | ECDSA-secp192r1-SHA256                               | ECDSA               | secp192r1                                                   | 192 bit    | SHA-256        | 48           |
-| ECDSA-secp256r1-SHA256 (default in OCMF Version 0.4) | ECDSA               | secp256r1, NIST P-256, ANSI X9.62 elliptic curve prime256v1 | 256 bit    | SHA-256        | 64           |
```

Comment:
For me, the previous and your updated text is confusing: does it mean that the default value is bound to a specific document version - if so then it might be good to document it for each released version. I don't have access to or know about versions prior to what is tracked here in git. I'd assume that it is ok to just use `ECDSA-secp256r1-SHA256 (default)` here and to drop the version reference completely.

Reply:
The most obvious interpreation is, that ***since*** OCMF v0.4 `secp256r1` is the default.

When ever there is another line like e.g. "ECDSA-secp521r1-SHA512 (default in OCMF Version 2.0)" then this is the new default since this version number.

Obviously replacing _"in"_ with _"since"_ would be a good idea.

I would be more concerned, that this document specifies to use SHA256 for ECC algorithms having more than 256 bits. This is reduces their security dramatically and will most likely lead to imcompatibilities with all crypto libs and TPMs that do not allow to specify a seperate hash algorithm for a signature algorithm.

Reply:
Good point, that the default shouldn't be version dependent. 

I don't see how the security is weakened by inputting 256 bits into a 384 curve, although the compatibility aspect definitely seems concerning. Maybe you could elaborate on this @ahzf

Reply:
Just ask you favorit AI ;)

> Using SHA-256 with the secp384r1 elliptic curve, while technically feasible, can have implications on the overall security of the cryptographic system.
> 
> Security Mismatch: The primary concern is the mismatch between the security levels of the hash algorithm and the elliptic curve. secp384r1 is designed to offer security comparable to a 192-bit symmetric key algorithm, which is higher than the 128-bit security level provided by SHA-256. This mismatch means that the hash function could become the weakest link in the cryptographic process, potentially reducing the overall security of the system to the security level of the hash algorithm.
> 
> Implementation Considerations: Standard implementations and protocols that use secp384r1 may expect SHA-384 by default. Using SHA-256 could require custom configurations or coding, which might introduce additional complexity or even vulnerabilities if not done carefully.
> 
> Interoperability Issues: If you are operating in an environment with predefined standards or interacting with external systems, using a non-standard hash algorithm like SHA-256 with secp384r1 might lead to compatibility issues.
> 
> Regulatory and Compliance Aspects: In some cases, regulatory or compliance requirements might dictate the use of specific algorithmic combinations. Deviating from these standards (like using SHA-256 instead of SHA-384 with secp384r1) could lead to non-compliance.
> 
> Performance Impact: The performance impact is usually minimal in the context of choosing between SHA-256 and SHA-384. However, it's worth noting that SHA-384 might be slightly slower due to its larger output size, but this is often negligible in most applications.
> 
> In summary, while using SHA-256 with secp384r1 is possible, it's important to consider the potential security, interoperability, and compliance implications of deviating from the standard pairing of secp384r1 with SHA-384.

Reply:
I have given this a bit more thought. While in theory it is correct that the difficulty to find a hash collision for the payload is at 128 bit, the difficulty to find a valid payload as hash collision is much more than 128 bit. 
Just a small extrapolation: We have more than 12 fields that have a cardinality of at least 1. Each of this needs more than 4 chars fixed, so the difficulty increases by at least  (8 (bits of char) * 4 * 12) / 2 (birthday paradox) = 96 bits,  which is much more than the 64 bits security difference of the secp384r1 curve to SHA-256.

Reply:
Why do you think you could simply add 96 Bits to 128 Bits? There are the 128 Bits from the SHA-256 constraints and 96 Bits from your argumentation about the payload constraints. And you argue to combine both to 224 Bits of security. However, this isn't how crypto security level work. The difficulty of finding a valid payload that hashes to a specific value (pre-image resistance) and the difficulty of finding any two payloads that hash to the same value (collision resistance) are separate issues.

Also your argumentation does not even cover the problems with interoperability. But this is the real problem: OCMF makes false claims about the security of secp384r1, when it internally just uses SHA-256. As charging stations are critical infrastructure and OCMF covers security of metrological measurements I would even argue, that this is a mandatory reportable cyber incident and should be reported to the BSI and PTB. Also all current versions of the OCMF "specification" have to be revoked and every vendors Eichrechtszertifizierung using secp384r1 with SHA-256 has to be revoked.

Reply:
You are right, security still will only be 128 bit. My reasoning was wrong. 

However I don't think this is that much of a problem. 128 bit of hash security with secp384r1 is just the same as security as SHA256 secp256k1. OCMF never claims that secpk384r1 with SHA256 has 192 bits of security.
Up until about ~2017 this was also allowed practice for SSL certificates and 128 bits of security is still considered reasonably secure.

Reply:
You are right, my reasoning was wrong. 
But the relevant attack here: A second pre image attack on SHA-256 is 256 bit hard, which is higher than the 192 bit security of secp384k1 as far as I see?

Reply:
Which kind of benefit has this discussion? I'm not a cryptograph and you are not a cryptograph. Guessing around about cryptography is not a very enlightening conversation.

Just accept the mistake within the OCMF specification, fix it, apologise and we are done.

Reply:
lmao, just drop in a bit of whataboutism. 
In the context of you, advocated for reporting this to the BSI and revoking all Eichrechtszertifizierungen just one message ago, your new message almost seems comical. 
While I am in favor of deprecating or removing those curve-hash pairs in favor of interoperability and standardization in upcoming versions of OCMF, I do not think those pairs pose an actual security issue in the context of OCMF.
The 128 bits of security your favorite AI previously dropped in are relevant in the context of collision attacks on SHA-256, but this attack doesn't make sense in the context of OCMF, as the Wallbox itself would have to perform it. but it is trusted in the whole Eichrechts-security model qua certification. All other relevant attacks have a generic security level of 256 bits.
If you would like to continue this discussion on a technical level I am fully down for it, but please don't resort to ad hominem arguments.
Maybe also a member of S.A.F.E. e.V. who is a "cryptograph" could comment on this.

Reply:
Well... when taking cryptography and e-mobilitiy seriously, you could have accepted the facts, fixed the issue and you would already have notified the BSI and PTB about this major security issue as the security level is at least far less than the "advertised" security level.

But YOU have chooen to bullshit around with strange arguments whether this is a bigger or not so big security issue. So do not start blaming me for your faults.

Reply:
First of all, I am not a member of S.A.F.E. e.V. nor am I affiliated to them in any way. Therefore it isn't my responsibility to fix the alleged security issue nor to report it to the BSI nor my fault that it got introduced.
Secondly, I still do not think, that this poses any security deterioration of the security in the context of OCMF, as collision attacks are not relevant and therefore the security of SHA-256 still is 256 bits.
I invite you to counter this argument on a technical level.

PS: True, maybe I am from an extraterrestrial civilization, but who knows that you aren't either ;)

Reply:
When you just sign a "0" and a "1" using secp384r1 you will just get two long versions of a "0" and a "1". Then secp384r1 ist "broken" because you use it in a wrong way.

The reason we use hashing algorithms in combination with signature algorithms is to transform a "normal" text into a text of fixed block size and higher entropy. Also some attacks get harder when you have to fight both.

Yet this is still not enough. As people misunderstand cryptography all the time we use an additional large random number in modern algorithms to avoid that the hash value is the same for the same user input.

So again you compare apples and oranges.

...

Not the letters... the quality of the content and the fact, that I do not hide my identity, nor sell my soul to the highest bidder makes me a trustworthy human ;)

Reply:
Totally true! This means you have still have +- 256 bits to bruteforce in the case of SHA-256, which definitely isn't feasible. So the weakest link here still are the 192 bits for bruteforcing the secp384r1 private key. 

Could you give me a link to a paper or something else, which attacks get easier, in this context?

I do not see, why the usage of nonces in ECDSA is relevant in this discussion?

Reply:
Again wrong... 2^256 operations would be needed for a brute-force preimage attack of SHA256, but just 2^128 operations for a collision attack due to the birthday paradox.

There is just a damn good reason why secp384r1 is standardized with SHA384 and there is 0.0 probability, that you will find a way around it.

For consultancy services you can book me, yes. But you would not accept my arguments, not the arguments from papers anyway, or?

Reply:
Sure, 128 bits for collision attacks. But I see no attack scenario where collision are relevant in the OCMF attack model as written above. For e.g. TLS this is a different story, but here I just do not see an attack making use of collisions that isn't initiated by a charging station (...which is trusted qua PTB certification).

Reply:
You always change the argumentation, when I show your previous argumentation was wrong. Isn't that boring?  ;)

PTB does not do a security audit. It just checks the most obvious things in order to get authentic measurement values.
Esp. all the crazy companies who take their entire charging station firmware to the PTB can not be trusted at all.

As soon as the OCMF data record has left the charging station your argument is invalid anyway. The ev driver does not optain the OCMF from the charging station directly.

OCMF does not have very strict input validation. There are also many free text fields which all offer the possibility of collissions via padding attacks.

But the main attacker is still you, the guy who tries hard to ignore a security issue ;)

Reply:
All I have written, I have written previously but I seem to be reiterating my arguments.

Let's assume the PTB audit cannot be trusted: Then the whole point of signing meter values is flawed, as the "malicious" wallbox could just sign every meter value it wants, rendering this whole stuff completely useless. Therefore we are forced to assume the wallbox is trusted (which I do not like, but this is how it apparently works).

Next the wallbox transfers the signed meter values out via OCPP or something else and a malicious attacker like an EMSP or something else could start an attack.

What would an attack try to achieve? Get data displayed as verified in the Transparenzsoftware, although it was modified.
How does he do this?
A) Try to forge the data, while maintaining the signature. => Crack SHA-256 (2nd preimage attack) => 256 bits of security
B) Try to forge the data, while maintaining a valid signature by the wallbox. => Crack Wallbox Private Key => 192 bits of security.

I do not see myself here as countering security. Even if your argumentation was correct, we would still have 128 bits of security, which is way enough.

Reply:
You are still just bullshitting around!

The PTB "certification" is already be flawed, as they are "not allowed" to complain about unsecure methods to obtain the user identification e.g. via RFID UIDs. So its security goal is flawed anyway.

The entire process of validating the measurement signatures is flawed as the ev driver receives the public key parallel to the signed meter values. So an internal attacker would just replace both. All ways to verify the public key are broken by design or just no one cares about.

All I'm arguing is that secp384r1 within your list is not secp384r1 as well-known to everyone, but a weaker version, as you fucked up the SHA384 requirement of the official secp384r1 standard. Whether or not secp384r1+SAH256 would be "secure enough" is not the question here.

Reply:
I am not bullshitting around, I try to assess wether your objection on the different curve parameters actually poses a security issues that is worth of "reporting to the BSI, PTB, revoking all Eichrechtszertifizierungen that use it and all previous OCMF versions" or if the protocol is still secure. You last comment seems more like bullshitting around to me.

I still do not see that the curve pair has a relevantly lower security level (i.e. lower than 192 bits), than one using SHA384 in the OCMF context, but you seem to ignore my arguments and rant about the bad state of IT Security in electromobility instead (which I partially agree with).

The list is still not mine, I did not contribute the curve pair, I am not a maintainer or any of the stakeholders in the protocol.

I will leave this as it is and invite @mhei to draw his own conclusions of this discussion, although I too am in favor of removing/deprecating the pair, due to seldom use.

Reply:
"An approved hash function, as specified in FIPS 180, shall be used during the generation of
digital signatures. The security strength associated with the ECDSA digital signature process is
no greater than the minimum of the security strength associated with the bit length of n and the
security strength of the hash function that is employed. Both the security strength of the hash function used and the security strength associated with the bit length of n shall meet or exceed
the security strength required for the digital signature process. The security strengths for the
ranges of the bit lengths of n and for each hash function is provided in SP 800-57. "

https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf
https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-57pt1r5.pdf

The security strength of the digital signature process is limited by the lower of the two: the security strength associated with the bit length of n (i.e., the ECC key) and the security strength of the hash function used.
- For secp384r1, the security strength associated with the bit length of n is 192 bits.
- The security strength of SHA-256 is effectively 128 bits against collision attacks.
- When SHA-256 is used with secp384r1 in the ECDSA digital signature process, the overall security strength of the signature is determined by the weaker of the two, which in this case is SHA-256's 128-bit collision resistance.

So we have a significant security issue here that OCMF secp384r1 is far less secure than normal secp384r1. Your security anecdotes are really not relevant.

...

ISO15118-20 even uses secp521r1 / SHA512... just try to harass Marc Mültin over your very special crypto anecdotes ;)

Reply:
Alright, you got me back to writing, but nice to see that we are back on a technical level.
As I already said twice, I do not see where a collision attack is an attack vector here. Maybe you can present me with a scenario?